ENIGMAsystem/SHELL/Widget_Systems/TinyFileDialogs/dialogs.cpp

```diff
@@ -0,0 +1,469 @@
+#include <stdlib.h>
+#include <string.h>
+#include <algorithm>
+#include <sstream>
+#include <vector>
+#include <string>
+
+using std::string;
+
+namespace enigma
+{
+    bool widget_system_initialize()
+    {
+        return true;
+    }
+}
+
+bool is_number(string input)
+{
+    if (input[0] >= '0' && input[0] <= '9')
+        return true;
+
+    if (input[0] == '+' && (input[1] >= '0' && input[1] <= '9'))
+        return true;
+
+    if (input[0] == '-' && (input[1] >= '0' && input[1] <= '9'))
+        return true;
+
+    return false;
+}
+
+class FileFilter {
+  string filter_buf;
+  std::vector<const char*> descriptions_;
+  std::vector<std::vector<const char*>> patterns_;
+  std::vector<const char* const*> cpatterns_;
+  std::vector<int> pattern_counts_;
+
+ public:
+  FileFilter(const string &filter): filter_buf(filter + "|") {
+    if (!filter.empty()) {
+      size_t start = 0;
+      std::vector<const char*> *curfilter = nullptr;
+      for (size_t i = 0; i < filter_buf.length(); ++i) {
+        if (filter_buf[i] == '|') {
+          filter_buf[i] = 0;
+          if (curfilter) {
+            curfilter->push_back(filter_buf.c_str() + start);
+            curfilter = nullptr;
+          } else {
+            descriptions_.push_back(filter_buf.c_str() + start);
+            patterns_.push_back({});
+            curfilter = &patterns_.back();
+          }
+          start = i + 1;
+        } else if (curfilter && filter_buf[i] == ';') {
+          filter_buf[i] = 0;
+          curfilter->push_back(filter_buf.c_str() + start);
+          start = i + 1;
+        }
+      }
+      if (descriptions_.size() > patterns_.size()) {
+        descriptions_.pop_back();
+      }
+    }
+    pattern_counts_.reserve(descriptions_.size());
+    for (auto &pv : patterns_) {
+      pattern_counts_.push_back(pv.size());
+      pv.push_back(nullptr);
+      cpatterns_.push_back(pv.data());
+    }
+    cpatterns_.push_back(nullptr);
+    descriptions_.push_back(nullptr);
+  }
+
+  size_t count() const { return patterns_.size(); }
+  const char* const* descriptions() { return descriptions_.data(); }
+  const char* const* const* patterns() { return cpatterns_.data(); }
+  const int* pattern_counts() { return pattern_counts_.data(); }
+};
+
+void show_error(string errortext, const bool fatal)
+{
+    string msg;
+
+    if (errortext == "")
+        msg = " ";
+    else
+        msg = errortext;
+
+    if (msg != " ")
+        msg = msg + "\n\n";
+
+    replace(msg.begin(), msg.end(), '"', '\'');
+
+    if (fatal == 0)
+    {
+        msg = msg + "Do you want to abort the application?";
+
+        double input = tinyfd_messageBox("Error", msg.c_str(), "yesno", "error", 1);
+
+        if (input == 1)
+            exit(0);
+    }
+    else
+    {
+        msg = msg + "Click 'OK' to abort the application.";
+
+        tinyfd_messageBox("Error", msg.c_str(), "ok", "error", 1);
+
+        exit(0);
+    }
+}
+
+namespace enigma_user
+{
+    char result[PATH_MAX];
```

Comment:
Please don't put this in `enigma_user`. Actually, you shouldn't need this at all.